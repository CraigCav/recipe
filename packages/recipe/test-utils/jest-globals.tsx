import React from 'react';
import {render, cleanup} from '@testing-library/react';
import {ThemeProvider} from 'emotion-theming';
import {standard} from '../themes';

import '@testing-library/jest-dom/extend-expect';

const renderWithTheme = renderFn => (component, ...rest) =>
  renderFn(<ThemeProvider theme={standard}>{component}</ThemeProvider>, rest);

export const renderToHtml = renderWithTheme((ui, options) => {
  const {container} = render(ui, options);
  return container.outerHTML;
});

export const fullRender = renderWithTheme(render);

afterEach(cleanup);
