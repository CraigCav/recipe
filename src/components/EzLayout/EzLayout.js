import PropTypes from 'prop-types';
import styled from 'react-emotion';
import {base, layout, spacing} from './EzLayout.styles';
import {standard} from '../../themes';

const layoutTypes = PropTypes.oneOf(['basic', 'right', 'equal', 'split', 'stack']);

/**
 * Layout provide common ways to arrange content in a single horizontal row.
 */
const EzLayout = styled('div')(base, layout, spacing);

EzLayout.propTypes = {
  /**
   * Controls the arrangement of items within the layout.
   */
  layout: PropTypes.oneOfType([layoutTypes, PropTypes.shape({base: layoutTypes})]),
  /**
   * The theme controlling the default styles.
   */
  theme: PropTypes.shape({
    colors: PropTypes.object,
  }),
};

/**
 * defaultProps
 * @property {string} layout - uses the basic layout by default.
 * @property {object} theme - uses the standard theme by default.
 */
EzLayout.defaultProps = {
  layout: 'basic',
  theme: standard,
};

/**
 * @component
 */
export default EzLayout;
