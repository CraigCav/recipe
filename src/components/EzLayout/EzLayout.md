---
name: EzLayout
title: Layout
path: "/components/ez-layout"
---

The layout component provide common ways to arrange the content of a screen. The layout component controls the spacing of its content. For horizontal layout types, the layout component will vertically center the content items.

---

## Best Practices

Layouts should:

* Be used for small-scale layout tasks
* Be used when you want to lay out a horizontal row of content OR
* Be used when you want to layout out a vertical stack of content that would otherwise be inline

Layouts should not:

* Be used for large-scale page layout. For large-scale layouts, consider using a [Card](/components/ez-card) within [Page Content](/components/ez-page-content)
* Wrap content onto multiple lines. The content inside a layout should be designed as not to exceed the available horizontal space.

---

## Examples

### Basic layout

The `basic` layout arranges content into a single row with even spacing between the content items.

```jsx
<EzCard
  className={css(`
  .basic > * {
    background-color: yellow;
  }
`)}
>
  <EzLayout layout="basic" className="basic">
    <div>Content</div>
    <div>Content</div>
    <div>Content</div>
    <div>Content</div>
  </EzLayout>
</EzCard>
```

### Right layout

The `right` layout arranges content into a single row, aligned to the right edge of its container, with even spacing between the content items. The `right` layout can be used, for example, for aligning action buttons within a card.

```jsx
<EzCard
  className={css(`
  .right > * {
    background-color: coral;
  }
`)}
>
  <EzLayout layout="right" className="right">
    <div>Content</div>
    <div>Content</div>
    <div>Content</div>
    <div>Content</div>
  </EzLayout>
</EzCard>
```

### Equal layout

The `equal` layout arranges content into equal widths filling a single row. The `equal` layout can be used, for example, for evenly spacing spending statistics across a single row.

```jsx
<EzCard
  className={css(`
  .equal > * {
    background-color: aqua;
  }
`)}
>
  <EzCardSection>
    <EzLayout layout="equal" className="equal">
      <div>Equal Layout</div>
      <div>Content</div>
      <div>Content</div>
    </EzLayout>
  </EzCardSection>
  <EzCardSection>
    <EzLayout layout="equal" className="equal">
      <div>Equal Layout</div>
      <div>Content</div>
      <div>Content</div>
      <div>Content</div>
    </EzLayout>
  </EzCardSection>
</EzCard>
```

### Split layout

The `split` layout arranges content into two groups on a single row, one aligned to the left edge of its container with the other to the right edge. The `split` layout will usually be [combined with other layout arrangements inside](#combining-layouts). The `split` layout can be used, for example, for aligning secondary buttons for next steps to the bottom left of a card, while aligning primary calls-to-action to the bottom right of a card.

```jsx
<EzCard
  className={css(`
  .split > * {
    background-color: palegreen;
  }
`)}
>
  <EzLayout layout="split" className="split">
    <div>Split Layout</div>
    <div>Content</div>
  </EzLayout>
</EzCard>
```

### Stack layout

The `stack` layout arranges content into a vertical column. The `stack` layout can be used, for example, when targeting small screen sizes, to organize content into stacked columns. In many cases, a `stack` layout may not be necessary; block level components will naturally stack in a container, and container components such as [`EzCard`](/components/ez-card) and [`EzPageContent`](/components/ez-page-content) will apply consistent spacing between content items.

```jsx
<EzCard
  className={css(`
  .stack > * {
    background-color: lightgray;
  }
`)}
>
  <EzLayout layout="stack" className="stack">
    <div>Stack Layout</div>
    <div>Content</div>
  </EzLayout>
</EzCard>
```

### Combining Layouts

The layout component can be nested in order to provide more complex or unique arrangements of components. For example, a toolbar might use a basic layout inside of an equal layout, to arrange logical groupings of action buttons into visual groups with even spacing between the grouped buttons.

```jsx
<EzCard
  className={css(`
  .basic > * {
    background-color: yellow;
  }
  .equal > * {
    background-color: aqua;
  }
`)}
>
  <EzCardSection>
    <EzLayout layout="split">
      <EzLayout layout="basic" className="basic">
        <div>Basic Layouts in a Split Layout</div>
        <div>Content</div>
      </EzLayout>
      <EzLayout layout="basic" className="basic">
        <div>Content</div>
        <div>Content</div>
      </EzLayout>
    </EzLayout>
  </EzCardSection>
  <EzCardSection>
    <EzLayout layout="split">
      <EzLayout layout="equal" className="equal">
        <div>Equal Layouts in a Split Layout</div>
        <div>Content</div>
      </EzLayout>
      <EzLayout layout="equal" className="equal">
        <div>Content</div>
        <div>Content</div>
      </EzLayout>
    </EzLayout>
  </EzCardSection>
</EzCard>
```

### Responsive Layouts

The layout component can be used to optimize the layout of content to suit various device sizes. For example, a group of action buttons may use the `stack` layout on small screens where horizontal space limited, and the `basic` layout on larger screens.

If you provide more than one layout, you should think of it as progressive enhancement. The `base` layout will be applied from the smallest devices up, until reaching the other layout. Then that second layout will takeover for larger devices. You can control where the transition happens.

While it is technically possible to pass more than two layouts (e.g. have a different layout each for phone, tablet, desktop) we don't see a need for that at this time. If designers run into a case for this kind of behavior they should bring it up for discussion.

```jsx
<EzCard
  className={css(`
  .responsive > * {
    background-color: lightgreen;
  }
`)}
>
  <EzLayout
    layout={{
      base: 'stack',
      medium: 'basic',
    }}
    className="responsive"
  >
    <div>Stack Layout</div>
    <div>Content</div>
  </EzLayout>
</EzCard>
```

---

## Related components

* [Card](/components/ez-card)
* [Page Content](/components/ez-page-content)
