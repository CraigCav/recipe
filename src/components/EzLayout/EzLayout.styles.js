import {css} from 'react-emotion';
import {margin} from 'polished';
import {responsive} from '../../styles';

export const base = css`
  display: flex;
`;

const reset = css`
  align-items: normal;
  flex-direction: row;
  justify-content: normal;

  > * {
    flex-basis: auto;
    flex-grow: 0;
  }
`;

export const basic = () => css`
  align-items: center;
  justify-content: flex-start;
`;

export const right = css`
  align-items: center;
  justify-content: flex-end;
`;

export const equal = () => css`
  align-items: center;
  justify-content: space-between;

  > * {
    flex-basis: 0;
    flex-grow: 1;
  }
`;

export const split = () => css`
  align-items: center;
  justify-content: space-between;
`;

export const stack = css`
  flex-direction: column;
`;

export const layout = responsive('layout', {
  basic,
  right,
  equal,
  split,
  stack,
  reset,
});

const setSpacing = (top, left) => css`
  > *:not(:first-child) {
    ${margin(top, null, null, left)};
  }
`;
const leftSpacing = ({theme}) => setSpacing(null, theme.spacing.sm);
const topSpacing = ({theme}) => setSpacing(theme.spacing.sm, null);

export const spacing = responsive('layout', {
  basic: leftSpacing,
  right: leftSpacing,
  equal: leftSpacing,
  stack: topSpacing,
  reset: setSpacing(0, 0),
});
