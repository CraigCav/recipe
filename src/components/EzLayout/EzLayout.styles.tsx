import {css} from '@emotion/core';
import {responsive} from '../../styles';

export const base = () => css`
  display: flex;
`;

const reset = () => css`
  align-items: normal;
  flex-direction: row;
  justify-content: normal;
  flex-wrap: nowrap;

  > * {
    flex-basis: auto;
    flex-grow: 0;
    width: auto;
  }
`;

export const basic = () => css`
  align-items: center;
  justify-content: flex-start;
`;

export const right = () => css`
  align-items: center;
  justify-content: flex-end;
`;

export const equal = () => css`
  align-items: center;
  justify-content: space-between;

  > * {
    flex-basis: 0;
    flex-grow: 1;
  }
`;

export const split = () => css`
  align-items: center;
  justify-content: space-between;
`;

export const stack = () => css`
  flex-direction: column;
`;

const columnWidth = ({columns, theme}) => {
  const width = n => ({width: `calc(${n < 2 ? 100 : 100 / n}% - ${theme.spacing.sm})`});

  if (typeof columns !== 'object') return width(columns);

  return responsive(
    'columns',
    Object.keys(columns).reduce((res, key) => {
      res[columns[key]] = width(columns[key]);
      return res;
    }, {})
  )({columns, theme});
};

const tile = ({columns, theme}) => css`
  flex-wrap: wrap;

  /* multiply by -1 to negate the margin, since the container absorbs the additional outer margin of the children) */
  margin: calc(${theme.spacing.sm} / 2 * -1);

  > * {
    flex-basis: auto;
    flex-grow: 0;
    ${columnWidth({columns, theme})};
    /* ↓ half the value, to avoid doubling up the space between columns */
    margin: calc(${theme.spacing.sm} / 2);
  }
`;

const cluster = ({theme}) => css`
  flex-wrap: wrap;
  justify-content: flex-start;
  align-items: center;

  /* multiply by -1 to negate the margin, since the container absorbs the additional outer margin of the children) */
  margin: calc(${theme.spacing.sm} / 2 * -1);

  > * {
    /* ↓ half the value, to avoid doubling up the space between columns */
    margin: calc(${theme.spacing.sm} / 2);
  }
`;

export const layout = responsive('layout', {
  basic,
  right,
  equal,
  split,
  stack,
  tile,
  cluster,
  reset,
});

const setSpacing = (top, left) => css`
  && > * + * {
    margin-top: ${top};
    margin-left: ${left};
  }
`;
const leftSpacing = ({theme}) => setSpacing(null, theme.spacing.sm);
const topSpacing = ({theme}) => setSpacing(theme.spacing.sm, null);

export const spacing = responsive('layout', {
  basic: leftSpacing,
  right: leftSpacing,
  equal: leftSpacing,
  stack: topSpacing,
  reset: () => ({'&& > * + *': {marginLeft: 0, marginRight: 0, marginTop: 0, marginBottom: 0}}),
});

export const wrapper = () => css`
  /* ↓ Suppress horizontal scrolling caused by the negative margin in some circumstances */
  overflow: hidden;
`;
