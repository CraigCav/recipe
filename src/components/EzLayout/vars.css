:root {
  --recipe-layout-gap: var(--recipe-global-static-size-150);
}