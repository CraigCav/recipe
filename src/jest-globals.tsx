import {shallow as enzymeShallow, render as enzymeRender, mount as enzymeMount} from 'enzyme';
import React from 'react';
import {create as reactCreate, ReactTestRenderer} from 'react-test-renderer';
import {render as rtlRender} from 'react-testing-library';
import {renderToStaticMarkup} from 'react-dom/server';

import {ThemeProvider} from 'emotion-theming';
import {standard} from './themes';

const renderWithTheme = renderFn => (component, ...rest) =>
  renderFn(<ThemeProvider theme={standard}>{component}</ThemeProvider>, rest);

export const shallow = tree => {
  const wrapper = enzymeShallow(<ThemeProvider theme={standard} />);
  const instance = wrapper.instance() as any;
  const context = instance.getChildContext();

  return enzymeShallow(tree, {context});
};

export const mount = tree => {
  const wrapper = enzymeShallow(<ThemeProvider theme={standard} />);
  const instance = wrapper.instance() as any;
  const context = instance.getChildContext();

  return enzymeMount(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes,
  });
};

export const renderToHtml = renderWithTheme(renderToStaticMarkup);

export const render = renderWithTheme(enzymeRender);

export const fullRender = renderWithTheme(rtlRender);

export const create = renderWithTheme(reactCreate);
