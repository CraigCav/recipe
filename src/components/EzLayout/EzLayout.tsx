import PropTypes from 'prop-types';
import {base, layout, spacing} from './EzLayout.styles';
import styled from '../../themes/styled';

const layoutTypes = PropTypes.oneOf(['basic', 'right', 'equal', 'split', 'stack']);

/**
 * Layout provide common ways to arrange content in a single horizontal row.
 */
const EzLayout = styled.div(base, layout, spacing);

EzLayout.propTypes = {
  /**
   * Controls the arrangement of items within the layout.
   */
  layout: PropTypes.oneOfType([layoutTypes, PropTypes.shape({base: layoutTypes})]),
};

/**
 * defaultProps
 * @property {string} layout - uses the basic layout by default.
 * @property {object} theme - uses the standard theme by default.
 */
EzLayout.defaultProps = {
  layout: 'basic',
};

/**
 * @component
 */
export default EzLayout;
