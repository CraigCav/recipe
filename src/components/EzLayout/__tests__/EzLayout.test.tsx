import React from 'react';
import EzLayout from '../EzLayout';
import {axe} from 'jest-axe';

import {create, renderToHtml} from '../../../jest-globals';

describe('EzLayout', () => {
  it('should render with default styles', () => {
    const actual = create(
      <EzLayout layout="basic">
        <div>Content</div>
        <div>Content</div>
      </EzLayout>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should render with right aligned content', () => {
    const actual = create(
      <EzLayout layout="right">
        <div>Content</div>
        <div>Content</div>
      </EzLayout>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should render with equally sized content', () => {
    const actual = create(
      <EzLayout layout="equal">
        <div>Content</div>
        <div>Content</div>
      </EzLayout>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should render with a split layout', () => {
    const actual = create(
      <EzLayout layout="split">
        <div>Content</div>
        <div>Content</div>
      </EzLayout>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should render with a stacked layout', () => {
    const actual = create(
      <EzLayout layout="stack">
        <div>Content</div>
        <div>Content</div>
      </EzLayout>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should render with a stacked layout on small screens, and equal layouts on larger screens', () => {
    const actual = create(
      <EzLayout layout={{base: 'stack', large: 'equal'}}>
        <div>Content</div>
        <div>Content</div>
      </EzLayout>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(
      <EzLayout layout="basic">
        <div>Content</div>
        <div>Content</div>
        <div>Content</div>
        <div>Content</div>
      </EzLayout>
    );
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
