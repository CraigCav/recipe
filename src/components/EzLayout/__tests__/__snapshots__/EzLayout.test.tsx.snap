// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EzLayout should render with a split layout 1`] = `
.recipe-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

<div
  className="recipe-0 recipe-1"
>
  <div>
    Content
  </div>
  <div>
    Content
  </div>
</div>
`;

exports[`EzLayout should render with a stacked layout 1`] = `
.recipe-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.recipe-0 > *:not(:first-child) {
  margin-top: 0.8571428571428571rem;
}

<div
  className="recipe-0 recipe-1"
>
  <div>
    Content
  </div>
  <div>
    Content
  </div>
</div>
`;

exports[`EzLayout should render with a stacked layout on small screens, and equal layouts on larger screens 1`] = `
.recipe-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media screen and (min-width:66.3125em) {
  .recipe-0 {
    -webkit-align-items: normal;
    -webkit-box-align: normal;
    -ms-flex-align: normal;
    align-items: normal;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-pack: normal;
    -webkit-justify-content: normal;
    -ms-flex-pack: normal;
    justify-content: normal;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
  }

  .recipe-0 > * {
    -webkit-flex-basis: auto;
    -ms-flex-preferred-size: auto;
    flex-basis: auto;
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
  }

  .recipe-0 > * {
    -webkit-flex-basis: 0;
    -ms-flex-preferred-size: 0;
    flex-basis: 0;
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
  }
}

.recipe-0 > *:not(:first-child) {
  margin-top: 0.8571428571428571rem;
}

@media screen and (min-width:66.3125em) {
  .recipe-0 > *:not(:first-child) {
    margin-top: 0;
    margin-left: 0;
  }

  .recipe-0 > *:not(:first-child) {
    margin-left: 0.8571428571428571rem;
  }
}

<div
  className="recipe-0 recipe-1"
>
  <div>
    Content
  </div>
  <div>
    Content
  </div>
</div>
`;

exports[`EzLayout should render with default styles 1`] = `
.recipe-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.recipe-0 > *:not(:first-child) {
  margin-left: 0.8571428571428571rem;
}

<div
  className="recipe-0 recipe-1"
>
  <div>
    Content
  </div>
  <div>
    Content
  </div>
</div>
`;

exports[`EzLayout should render with equally sized content 1`] = `
.recipe-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.recipe-0 > * {
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.recipe-0 > *:not(:first-child) {
  margin-left: 0.8571428571428571rem;
}

<div
  className="recipe-0 recipe-1"
>
  <div>
    Content
  </div>
  <div>
    Content
  </div>
</div>
`;

exports[`EzLayout should render with right aligned content 1`] = `
.recipe-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.recipe-0 > *:not(:first-child) {
  margin-left: 0.8571428571428571rem;
}

<div
  className="recipe-0 recipe-1"
>
  <div>
    Content
  </div>
  <div>
    Content
  </div>
</div>
`;
